lib/Service/AvatarService.php

```diff
@@ -0,0 +1,235 @@
+
+		$mimeType = $image->mimeType();
+		$allowedMimeTypes = [
+			'image/jpeg',
+			'image/png',
+		];
+		if (!in_array($mimeType, $allowedMimeTypes)) {
+			throw new InvalidArgumentException($this->l->t('Unknown filetype'));
+		}
+
+		$this->deleteAvatarFromFilesystem($room);
+
+		$token = $room->getToken();
+		$avatarFolder = $this->getAvatarFolder($token);
+		$avatarName = $this->getRandomAvatarName($room);
+		$avatarFolder->newFile($avatarName, $image->data());
+		$this->roomService->setAvatar($room, $avatarName);
+	}
+
+	private function getRandomAvatarName(Room $room): string {
+		$name = $this->random->generate(10, ISecureRandom::CHAR_HUMAN_READABLE);
+		if ($name === $room->getName()) {
+			return $this->getRandomAvatarName($room);
+		}
+		return $name;
+	}
+
+	private function getAvatarFolder(string $token): ISimpleFolder {
+		try {
+			$folder = $this->appData->getFolder('room-avatar');
+		} catch (NotFoundException $e) {
+			$folder = $this->appData->newFolder('room-avatar');
+		}
+		try {
+			$avatarFolder = $folder->getFolder($token);
+		} catch (NotFoundException $e) {
+			$avatarFolder = $folder->newFolder($token);
+		}
+		return $avatarFolder;
+	}
+
+	public function getAvatar(Room $room, ?IUser $user, bool $darkTheme = false): ISimpleFile {
+		$token = $room->getToken();
+		$avatar = $room->getAvatar();
+		if ($avatar) {
+			try {
+				$folder = $this->appData->getFolder('room-avatar');
+				if ($folder->fileExists($token)) {
+					$file = $folder->getFolder($token)->getFile($avatar);
+				}
+			} catch (NotFoundException $e) {
+			}
+		}
+		// Fallback
+		if (!isset($file)) {
+			if ($room->getType() === Room::TYPE_ONE_TO_ONE) {
+				$users = json_decode($room->getName(), true);
+				foreach ($users as $participantId) {
+					if ($user instanceof IUser && $participantId !== $user->getUID()) {
+						$avatar = $this->avatarManager->getAvatar($participantId);
+						$file = $avatar->getFile(512, $darkTheme);
+					}
+				}
+			} elseif ($room->getObjectType() === 'file') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-text-white.svg'));
+			} elseif ($room->getObjectType() === 'share:password') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-password-white.svg'));
+			} elseif ($room->getObjectType() === 'emails') {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-mail-white.svg'));
+			} elseif ($room->getType() === Room::TYPE_PUBLIC) {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-public-white.svg'));
+			} else {
+				$file = new InMemoryFile($token, file_get_contents(__DIR__ . '/../../img/icon-contacts-white.svg'));
+			}
+		}
+		return $file;
+	}
+
+	private function deleteAvatarFromFilesystem(Room $room): void {
+		$token = $room->getToken();
+		$avatarFolder = $this->getAvatarFolder($token);
+		if (count($avatarFolder->getDirectoryListing()) >= 1) {
+			$avatarFolder->delete();
```

Comment:
There is no need to remove the complete folder.
Could just loop over `getDirectoryListing` and delete all images which are not the new one?